File: comparison/generics/generics.hs
#!/usr/bin/env runghc

data Box a = Box { item :: a } deriving (Show)

newBox :: a -> Box a
newBox i = Box { item = i }


class Management b where
  get :: b a -> Either String a
  set :: b a -> a -> Either String ()

instance Management Box where
  get (Box i) = Right i
  set (Box _i) _v = Right ()


main :: IO ()
main = do
  let myBox = newBox 100

  case get myBox of
    Right i -> putStrLn $ "Item: " <> show i
    Left e -> putStrLn $ "Failed to get item: " <> e
  
  case set myBox 1000 of
    Right _ -> putStrLn "Item updated successfully"
    Left e -> putStrLn $ "Failed to set item: " <> e

File: comparison/generics/generics.rs
#!/usr/bin/env rust-script

struct MyBox<T> {
    item: T,
}

impl<T> MyBox<T> {
    fn new(item: T) -> Self {
        Self { item }
    }
}

trait Management<T> {
    fn get(&self) -> Result<T, String>;
    fn set(
        &mut self,
        value: T,
    ) -> Result<(), String>;
}

impl<T> Management<T> for MyBox<T>
where
    T: Clone,
{
    fn get(&self) -> Result<T, String> {
        Ok(self
            .item
            .clone())
    }

    fn set(
        &mut self,
        value: T,
    ) -> Result<(), String> {
        self.item = value;
        Ok(())
    }
}

fn main() -> Result<(), String> {
    let mut my_box = MyBox::new(100);

    match my_box.get() {
        | Ok(item) => println!("Item: {}", item),
        | Err(e) => println!("Failed to get item: {}", e),
    }

    match my_box.set(1000) {
        | Ok(_) => println!("Item updated successfully"),
        | Err(e) => println!("Failed to set item: {}", e),
    }

    match my_box.get() {
        | Ok(item) => println!("Item: {}", item),
        | Err(e) => println!("Failed to get item: {}", e),
    }

    Ok(())
}
